# Notebook 2 — Du document au chunk indexé, et la première mesure du banc

Ce notebook suit un document sur tout le chemin de l'**ingestion** : parsé en sections, découpé
en **chunks**, transformé en **embeddings**, rangé dans **pgvector**, retrouvé par une question,
assemblé dans un prompt, cité dans une réponse. Chaque étape affiche ce qu'elle produit — le
chunk exact, le vecteur, la requête SQL de recherche, le prompt final, la réponse brute et ses
citations. Puis on mesure : trois modèles d'embeddings sur SciFact, comparés aux scores publiés.

Tout est importé de `src/` (ADR 007) et exécuté tel quel. **Prérequis** : la base locale
(`make k3d-up` puis `make pg-port-forward`, guide infra 01) — en mode échantillon comme en mode
complet. Sans clé d'API, les embeddings et la génération sont simulés (un embedding « sac de
mots hachés » et un provider scripté, tous deux dans `src/llm_testbench/llm/testing.py`) ; avec
`.env`, ce sont les vrais modèles.

## Partie 1 — Le problème

Un modèle de langage ne connaît pas vos documents. La solution naïve — coller le document entier
dans le prompt — casse vite : la **fenêtre de contexte** (la quantité de texte qu'un modèle
accepte) est finie, chaque token en entrée se paie, et un modèle noyé dans cinquante pages
répond moins bien qu'un modèle qui reçoit les trois paragraphes utiles.

D'où le **RAG** (*retrieval-augmented generation*) : on cherche d'abord les passages pertinents,
on les donne au modèle, il répond en les citant. Pour chercher, il faut avoir préparé les
documents — c'est l'**ingestion**, et elle décide de la moitié de la qualité finale :

| Étape | Question qu'elle règle | Ce qui casse si on la bâcle |
|---|---|---|
| **Parsing** | comment un fichier devient du texte structuré | tableaux mélangés, notes de bas de page au milieu des phrases |
| **Chunking** | en morceaux de quelle taille, coupés où | une idée coupée en deux, ou un chunk si long que l'information utile s'y noie |
| **Embeddings** | comment représenter un chunk pour le retrouver | un modèle qui rapproche des textes qui n'ont rien à voir |
| **Index** | où ranger les vecteurs et comment chercher vite | une recherche linéaire qui ne tient plus à cent mille chunks |
| **Retrieval** | quels chunks pour cette question | trop peu (l'information manque), trop (le bruit gagne) |

Ce notebook construit la version **volontairement naïve** de chaque étape (la baseline que la
phase 3 devra battre, chiffres à l'appui) — sauf les embeddings, qu'on mesure dès maintenant.

In [1]:
# --- Configuration : modifiez, puis « Run All » ---
import os
from pathlib import Path

SAMPLE = os.environ.get("LLM_TESTBENCH_SAMPLE", "0") == "1"

ENV_FILE = Path.cwd().parent / ".env"
if ENV_FILE.is_file():
    for line in ENV_FILE.read_text().splitlines():
        if "=" in line and not line.startswith("#"):
            key, _, value = line.partition("=")
            os.environ.setdefault(key.strip(), value.strip())

DATABASE_URL = os.environ.get(
    "LLM_TESTBENCH_DATABASE_URL",
    "postgresql://testbench:testbench-local-only@localhost:5432/testbench",
)
# Vrais modèles seulement en mode complet ET avec les deux clés.
USE_REAL_API = (
    (not SAMPLE)
    and bool(os.environ.get("OPENAI_API_KEY"))
    and bool(os.environ.get("ANTHROPIC_API_KEY"))
)

EMBEDDING_MODEL = "openai/text-embedding-3-small"
EMBEDDING_DIMENSIONS = 1536 if USE_REAL_API else 256  # l'embedding simulé a 256 dimensions
CHAT_MODEL = "anthropic/claude-sonnet-5"

N_PAPERS = 2 if SAMPLE else 20  # articles QASPER ingérés
CHUNK_SIZE = 120 if SAMPLE else 800  # caractères (les fixtures sont minuscules)
CHUNK_OVERLAP = 30 if SAMPLE else 150
K = 4  # chunks récupérés par question
SHOW_PAPER = 0  # article détaillé dans les sorties
QUESTION_INDEX = 0  # question annotée de cet article, posée au système

INDEX_KEY = "nb2-sample" if SAMPLE else "qasper-nb2-oai-small"
SOURCE_KEY = "docs:qasper-nb2"
RUN_API_BENCH = False  # True : mesure aussi text-embedding-3-small (~2 min, ~0,03 $)

print(
    f"mode échantillon : {SAMPLE} — vrais modèles : {USE_REAL_API} — base : {DATABASE_URL.split('@')[-1]}"
)

mode échantillon : True — vrais modèles : False — base : localhost:5432/testbench


In [2]:
import inspect
import time
from itertools import pairwise

from IPython.display import Markdown, display
from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter

from llm_testbench.app.chat import ChatService
from llm_testbench.app.schemas import ChatRequest
from llm_testbench.config import Settings, find_repo_root
from llm_testbench.eval.embeddings_bench import PUBLISHED_SCIFACT, load_report
from llm_testbench.eval.loaders import QasperLoader
from llm_testbench.generation.answer import REFUSAL_SENTINEL, AnswerBuilder
from llm_testbench.ingest.chunking import RecursiveChunker
from llm_testbench.ingest.pipeline import IngestionPipeline
from llm_testbench.ingest.store import PgVectorStore
from llm_testbench.llm import DataClass, ModelRef
from llm_testbench.llm.cache import cosine_similarity
from llm_testbench.llm.client import LLMClient
from llm_testbench.llm.embeddings import EmbeddingRequest
from llm_testbench.llm.prompts import PromptRegistry
from llm_testbench.llm.testing import (
    HashingEmbedder,
    Step,
    anthropic_message,
    anthropic_sse,
    scripted_anthropic_provider,
)
from llm_testbench.obs.tracing import configure_tracing, get_tracer
from llm_testbench.sources import SearchQuery
from llm_testbench.sources.documents import DocumentSource

ROOT = find_repo_root()
settings = Settings(
    data_dir=ROOT / "data" / "cache",
    fixtures_dir=(ROOT / "tests" / "fixtures") if SAMPLE else None,
    sample=SAMPLE,
)


def show_source(*objects: object) -> None:
    """Affiche la source réelle d'objets de src/ (pur affichage)."""
    display(Markdown("\n".join(f"```python\n{inspect.getsource(o)}```" for o in objects)))


exporter = InMemorySpanExporter()
client = LLMClient.from_settings(
    settings, tracer=get_tracer(configure_tracing(exporter=exporter, set_global=False))
)
if not USE_REAL_API:
    client.register_embedder("openai", HashingEmbedder(dimensions=EMBEDDING_DIMENSIONS))
    scripted_provider, script = scripted_anthropic_provider(
        [
            Step(
                json_body=anthropic_message(
                    "D'après les extraits, la première étape est le retrieval [1]. Il précède la génération [2].",
                    input_tokens=700,
                    output_tokens=30,
                )
            ),
            Step(json_body=anthropic_message(REFUSAL_SENTINEL, input_tokens=650, output_tokens=18)),
            Step(sse_body=anthropic_sse(["Le retrieval ", "trouve les documents ", "[1]."])),
        ]
    )
    client.register_provider("anthropic", scripted_provider)

store = PgVectorStore.from_url(DATABASE_URL)
await store.open()
await store.ensure_schema()
print(
    f"racine : {ROOT} — embeddings : {'réels' if USE_REAL_API else 'simulés (sac de mots hachés)'} — base prête"
)

racine : /Users/christophe/Roadmaps/roadmap-ai-engineer — embeddings : simulés (sac de mots hachés) — base prête


## Partie 2 — Les options

| Décision | Options | Retenu en phase 1 (ADR 007), et pourquoi |
|---|---|---|
| Documents | passages pré-découpés (BEIR) / **articles complets** (QASPER) | QASPER : des documents longs, en sections, sans parser de PDF (texte intégral fourni) — le chunking y fait quelque chose |
| Chunking | fixe / **récursif par séparateurs** / sémantique / parent-document / *late chunking* | récursif : la baseline lisible en 60 lignes ; les autres sont la phase 3, mesurées contre elle |
| Embeddings | modèle local (bge, MiniLM…) / **API** (OpenAI) / Bedrock | API par défaut pour la simplicité ; trois modèles comparés en partie 5 |
| Stockage | pgvector / Qdrant, Weaviate / FAISS en mémoire | pgvector : une seule base, une table de vecteurs **par index**, HNSW cosinus — le registre d'index est le début du lineage de la phase 3 |
| Réponse | JSON structuré / **texte avec `[n]` et phrase-sentinelle de refus** | compatible avec le streaming, un seul parseur pour les deux transports |

## Partie 3 — Lecture guidée, étape par étape

### 3.1 Parsing : un article QASPER devient un document en sections

QASPER (Dasigi et al. 2021) fournit des articles scientifiques complets, déjà structurés en
sections et paragraphes, avec des questions annotées — c'est le jeu du chunking « sur documents
bruts » de la phase 3. Le loader (`eval/loaders/qasper.py`) produit des `ParsedDocument` : un
titre, des sections ordonnées, une provenance, une empreinte.

In [3]:
dataset = QasperLoader(settings).load("validation", max_papers=N_PAPERS)
print("jeu :", dataset.describe())

paper = dataset.papers[SHOW_PAPER]
document = paper.to_document()
print(
    f"\narticle : « {document.title} » — {document.char_count} caractères, {len(document.sections)} sections"
)
print(f"{'#':>3}  {'section':<40} {'caractères':>10}")
for section in document.sections:
    print(f"{section.order:>3}  {section.title[:40]:<40} {len(section.text):>10}")
print(f"\nquestions annotées sur cet article : {[q.question for q in paper.questions][:3]}")

jeu : {'dataset': 'qasper', 'split': 'validation', 'papers': 2, 'sections': 3, 'questions': 3, 'sampled': False}

article : « A Tiny Paper on Retrieval » — 145 caractères, 3 sections
  #  section                                  caractères
  0  Abstract                                         33
  1  Introduction                                     56
  2  Method                                           56

questions annotées sur cet article : ['What is the first step of RAG?', 'Which similarity is used?']


### 3.2 Chunking récursif : couper là où le texte se coupe déjà

Un **chunk** est un morceau de document indexé et retrouvé comme un tout. Trop long, il dilue
l'information utile et déborde la fenêtre du modèle d'embeddings (512 tokens pour bge, 8 191
pour text-embedding-3 — au-delà, **troncature silencieuse**). Trop court, il perd son contexte.

Le chunker récursif essaie de couper sur le séparateur le plus naturel (paragraphe), puis descend
d'un cran si un morceau reste trop long (ligne, phrase, mot, caractère), et regroupe jusqu'à
`chunk_size` caractères avec un **chevauchement** (`chunk_overlap`) pour qu'une idée coupée ait
un témoin des deux côtés. Lisez `_split` : c'est tout l'algorithme.

In [4]:
show_source(RecursiveChunker._split, RecursiveChunker.chunk)

chunker = RecursiveChunker(chunk_size=CHUNK_SIZE, chunk_overlap=CHUNK_OVERLAP)
chunks = chunker.chunk(document)
print(
    f"{len(chunks)} chunks pour {document.char_count} caractères (taille {CHUNK_SIZE}, chevauchement {CHUNK_OVERLAP})\n"
)
print(f"{'id':<14} {'section':<22} {'pos':>9}  début du texte")
for chunk in chunks[:12]:
    print(
        f"{chunk.chunk_id:<14} {chunk.section_title[:22]:<22} {chunk.start_char:>4}-{chunk.end_char:<4}  {chunk.text[:60]!r}"
    )

pairs = [(a, b) for a, b in pairwise(chunks) if a.section_order == b.section_order]
if pairs:
    a, b = pairs[0]
    print(f"\nchevauchement entre {a.chunk_id} et {b.chunk_id} (même section) :")
    print(f"  fin de {a.chunk_id}   : …{a.text[-70:]!r}")
    print(f"  début de {b.chunk_id} : {b.text[:70]!r}…")

```python
    def _split(self, text: str, separators: tuple[str, ...]) -> list[str]:
        if len(text) <= self.chunk_size:
            return [text]
        separator = next(s for s in separators if s == "" or s in text)
        remaining = separators[separators.index(separator) + 1 :]
        chunks: list[str] = []
        window: list[str] = []
        length = 0
        for piece in _split_keep(text, separator):
            if len(piece) > self.chunk_size:
                if window:
                    chunks.append("".join(window))
                    window, length = [], 0
                chunks.extend(self._split(piece, remaining))
                continue
            if window and length + len(piece) > self.chunk_size:
                chunks.append("".join(window))
                while window and length > self.chunk_overlap:
                    length -= len(window.pop(0))
            window.append(piece)
            length += len(piece)
        if window:
            chunks.append("".join(window))
        return chunks
```
```python
    def chunk(self, document: ParsedDocument) -> list[Chunk]:
        chunks: list[Chunk] = []
        ordinal = 0
        for section in document.sections:
            cursor = 0
            for piece in self.split_text(section.text):
                text = piece.strip()
                start = section.text.find(text, cursor)
                if start < 0:  # pragma: no cover - garde-fou, ne devrait jamais arriver
                    start = section.text.find(text)
                end = start + len(text)
                chunks.append(
                    Chunk(
                        chunk_id=f"{document.doc_id}#{ordinal:04d}",
                        doc_id=document.doc_id,
                        ordinal=ordinal,
                        text=text,
                        section_title=section.title,
                        section_order=section.order,
                        start_char=start,
                        end_char=end,
                        checksum=sha256_text(text),
                        token_estimate=estimate_tokens(text),
                    )
                )
                cursor = start + 1
                ordinal += 1
        return chunks
```

3 chunks pour 145 caractères (taille 120, chevauchement 30)

id             section                      pos  début du texte
1001.0001#0000 Abstract                  0-33    'We study retrieval. It is useful.'
1001.0001#0001 Introduction              0-56    'Retrieval finds documents.\n\nIt is the first step of RAG.'
1001.0001#0002 Method                    0-56    'We embed chunks.\n\nThen we search with cosine similarity.'


### 3.3 Embeddings : un vecteur par chunk, et une distance entre vecteurs

Un **embedding** est un vecteur de nombres qui représente le sens d'un texte : deux textes
proches par le sens ont des vecteurs proches. La **similarité cosinus** (1 = même direction,
0 = orthogonaux) est la distance qu'on utilise pour chercher.

Ici l'embedding passe par `LLMClient.embed` : même gouvernance que les complétions (le texte
part dehors), même coût calculé, même trace. En mode simulé, l'embedder est un « sac de mots
hachés » : rustique (aucune sémantique, seulement des mots communs), mais déterministe et hors
ligne — regardez la matrice de similarité, et gardez-la en tête pour la partie 5.

In [5]:
samples = [
    chunks[0].text,
    chunks[1].text if len(chunks) > 1 else chunks[0].text,
    "Une recette de crêpes : farine, œufs, lait.",
]
result = await client.embed(
    EmbeddingRequest(
        model=ModelRef.parse(EMBEDDING_MODEL),
        texts=tuple(samples),
        data_class=DataClass.PUBLIC,
        dimensions=EMBEDDING_DIMENSIONS,
    )
)
print(
    f"modèle : {result.model} — {result.dimensions} dimensions — usage : {result.usage.model_dump()} — coût : {result.cost_usd}"
)
nonzero = [(i, round(v, 3)) for i, v in enumerate(result.vectors[0]) if v][:8]
print(
    f"vecteur du chunk 0 : {sum(1 for v in result.vectors[0] if v)} composantes non nulles sur {result.dimensions}, ex. (index, valeur) :",
    nonzero,
    "…",
)
print("\nsimilarité cosinus :")
labels = ["chunk 0", "chunk 1", "recette"]
print(" " * 9 + "".join(f"{label:>9}" for label in labels))
for i, vi in enumerate(result.vectors):
    print(
        f"{labels[i]:>9}" + "".join(f"{cosine_similarity(vi, vj):>9.2f}" for vj in result.vectors)
    )

modèle : hashing-256d — 256 dimensions — usage : {'input_tokens': 32, 'output_tokens': 0, 'cache_read_tokens': 0, 'cache_write_tokens': 0} — coût : None
vecteur du chunk 0 : 4 composantes non nulles sur 256, ex. (index, valeur) : [(53, -0.5), (193, 0.5), (202, -0.5), (235, 0.5)] …

similarité cosinus :
           chunk 0  chunk 1  recette
  chunk 0     1.00     0.18     0.00
  chunk 1     0.18     1.00    -0.32
  recette     0.00    -0.32     1.00


### 3.4 Le stockage : Postgres, une table de vecteurs par index

Le schéma (`ingest/sql/001_schema.sql`) sépare ce qui est partagé — `documents`, `chunks` — de
ce qui dépend d'un modèle d'embeddings : une table `emb_<clé>` **par index**, à dimension fixée,
avec un index **HNSW** en distance cosinus. Le registre `indexes` note, pour chaque index, le
chunker et ses paramètres, le modèle et sa dimension : de quoi régénérer n'importe quelle ligne
d'un futur tableau de résultats. La recherche est une seule requête SQL : lisez `search`.

In [6]:
print((ROOT / "src" / "llm_testbench" / "ingest" / "sql" / "001_schema.sql").read_text())
show_source(PgVectorStore.search)

-- Schéma de l'ingestion (ADR 007). Idempotent : rejoué à chaque démarrage.
--
-- documents et chunks sont partagés par tous les index d'une même source ; les vecteurs
-- vivent dans une table par index (emb_<clé>, créée par le code avec sa dimension), et
-- la table indexes est le registre qui relie chaque index au chunker et au modèle qui
-- l'ont produit — le début du lineage de la phase 3.

CREATE EXTENSION IF NOT EXISTS vector;

CREATE TABLE IF NOT EXISTS documents (
    source_key  text        NOT NULL,
    doc_id      text        NOT NULL,
    title       text        NOT NULL DEFAULT '',
    uri         text,
    checksum    text        NOT NULL,
    char_count  integer     NOT NULL,
    metadata    jsonb       NOT NULL DEFAULT '{}'::jsonb,
    ingested_at timestamptz NOT NULL DEFAULT now(),
    PRIMARY KEY (source_key, doc_id)
);

CREATE TABLE IF NOT EXISTS chunks (
    source_key     text    NOT NULL,
    chunk_id       text    NOT NULL,
    doc_id         text    NOT NULL,
  

```python
    async def search(
        self,
        spec: IndexSpec,
        vector: Sequence[float],
        *,
        k: int = 5,
        filters: dict[str, str] | None = None,
    ) -> list[SearchHit]:
        """Les ``k`` chunks les plus proches en cosinus, avec leur document. Les filtres
        s'appliquent aux métadonnées du document (inclusion JSON)."""
        needle = np.asarray(vector, dtype=np.float32)
        query = sql.SQL(
            "SELECT c.*, d.title AS document_title, d.uri AS document_uri,"
            " d.metadata AS document_metadata, d.ingested_at,"
            " 1 - (e.embedding <=> %(v)s) AS score"
            " FROM {} e"
            " JOIN chunks c ON c.source_key = e.source_key AND c.chunk_id = e.chunk_id"
            " JOIN documents d ON d.source_key = c.source_key AND d.doc_id = c.doc_id"
            " WHERE e.source_key = %(source)s AND d.metadata @> %(filters)s"
            " ORDER BY e.embedding <=> %(v)s LIMIT %(k)s"
        ).format(sql.Identifier(spec.table))
        params = {
            "v": needle,
            "source": spec.source_key,
            "filters": json.dumps(filters or {}),
            "k": k,
        }
        async with self._pool.connection() as conn, conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(query, params)
            rows = await cur.fetchall()
        return [
            SearchHit(
                chunk=_chunk_from_row(row),
                score=float(row["score"]),
                document_title=row["document_title"],
                document_uri=row["document_uri"],
                document_metadata=row["document_metadata"],
                ingested_at=row["ingested_at"],
            )
            for row in rows
        ]
```

### 3.5 Le pipeline : idempotent à deux niveaux

`IngestionPipeline.ingest` enchaîne parsing → chunking → embeddings par lots → écriture. Deux
protections contre le gaspillage : un document dont l'**empreinte** n'a pas changé n'est pas
réécrit ; un chunk qui a déjà son vecteur **dans cet index** n'est pas ré-embarqué. Relancez
l'ingestion : zéro embedding, zéro coût. C'est la propriété que la phase 3 industrialisera.

In [7]:
show_source(IngestionPipeline.ingest)

pipeline = IngestionPipeline(
    client=client, store=store, chunker=chunker, data_class=DataClass.PUBLIC, batch_size=32
)
spec = pipeline.index_spec(
    index_key=INDEX_KEY,
    source_key=SOURCE_KEY,
    embedding_model=EMBEDDING_MODEL,
    dimensions=EMBEDDING_DIMENSIONS,
    description=f"notebook 2 — QASPER × recursive({CHUNK_SIZE}/{CHUNK_OVERLAP})",
)
existing = await store.get_index(INDEX_KEY)
if existing is not None and existing != spec:
    print("l'index existe avec une autre définition (paramètres modifiés) : on le reconstruit")
    await store.drop_index(INDEX_KEY)
    await store.delete_source(SOURCE_KEY)

report = await pipeline.ingest(spec, dataset.documents())
print("\npremière ingestion :", report.model_dump(exclude={"index_key"}))
again = await pipeline.ingest(spec, dataset.documents())
print(
    "seconde ingestion  :",
    {
        k: v
        for k, v in again.model_dump().items()
        if k in ("documents_written", "chunks_embedded", "embedding_calls", "cost_usd")
    },
)
print("\nregistre :", (await store.get_index(INDEX_KEY)).model_dump())

```python
    async def ingest(self, spec: IndexSpec, documents: Iterable[ParsedDocument]) -> IngestReport:
        if spec.chunker != self._chunker.describe():
            raise ValueError(
                "le chunker du pipeline ne correspond pas à celui de l'index "
                f"({spec.chunker} ≠ {self._chunker.describe()})"
            )
        start = time.perf_counter()
        await self._store.register_index(spec)
        seen = written = chunks_total = embedded = tokens = calls = 0
        cost: float | None = 0.0
        pending: list[Chunk] = []

        async def flush() -> None:
            nonlocal embedded, tokens, calls, cost
            if not pending:
                return
            result = await self._client.embed(
                EmbeddingRequest(
                    model=spec.embedding_model,
                    texts=tuple(c.text for c in pending),
                    data_class=self._data_class,
                    dimensions=spec.dimensions,
                )
            )
            await self._store.add_embeddings(spec, list(zip(pending, result.vectors, strict=True)))
            embedded += len(pending)
            tokens += result.usage.input_tokens
            calls += 1
            cost = None if cost is None or result.cost_usd is None else cost + result.cost_usd
            pending.clear()

        for document in documents:
            seen += 1
            chunks = self._chunker.chunk(document)
            chunks_total += len(chunks)
            if await self._store.upsert_document(spec.source_key, document, chunks):
                written += 1
            pending.extend(await self._store.missing_embeddings(spec, chunks))
            while len(pending) >= self._batch_size:
                batch, rest = pending[: self._batch_size], pending[self._batch_size :]
                pending[:] = batch
                await flush()
                pending.extend(rest)
        await flush()

        return IngestReport(
            index_key=spec.index_key,
            documents_seen=seen,
            documents_written=written,
            chunks_total=chunks_total,
            chunks_embedded=embedded,
            embedding_tokens=tokens,
            embedding_calls=calls,
            cost_usd=cost,
            duration_ms=(time.perf_counter() - start) * 1000.0,
        )
```


première ingestion : {'documents_seen': 2, 'documents_written': 2, 'chunks_total': 5, 'chunks_embedded': 5, 'embedding_tokens': 50, 'embedding_calls': 1, 'cost_usd': None, 'duration_ms': 42.39583294838667}
seconde ingestion  : {'documents_written': 0, 'chunks_embedded': 0, 'embedding_calls': 0, 'cost_usd': 0.0}

registre : {'index_key': 'nb2-sample', 'source_key': 'docs:qasper-nb2', 'embedding_model': {'backend': 'openai', 'model': 'text-embedding-3-small'}, 'dimensions': 256, 'chunker': {'chunker': 'recursive', 'chunk_size': 120, 'separators': "('\\n\\n', '\\n', '. ', ' ', '')", 'chunk_overlap': 30}, 'description': 'notebook 2 — QASPER × recursive(120/30)'}


## Partie 4 — Exécution : une question, du retrieval à la réponse citée

### 4.1 Retrieval : la question devient un vecteur, les chunks les plus proches deviennent des preuves

La question est une **question annotée du jeu** (les auteurs de QASPER l'ont posée sur cet
article). `DocumentSource.search` l'embarque avec le même modèle que les chunks — obligatoire :
deux modèles différents donnent des vecteurs incomparables — puis interroge pgvector. Chaque
résultat est une `Evidence` : texte, score, rang et **provenance** complète.

In [8]:
show_source(DocumentSource.search)

source = DocumentSource(client=client, store=store, spec=spec)
question = paper.questions[QUESTION_INDEX].question
print(f"question : {question!r}\n")
evidence = await source.search(SearchQuery(text=question, k=K, data_class=DataClass.PUBLIC))
print(f"{'rang':>4} {'score':>6}  {'document':<12} {'localisateur':<44} texte")
for item in evidence:
    print(
        f"{item.rank:>4} {item.score:>6.3f}  {item.provenance.document_id:<12} {item.provenance.locator[:44]:<44} {item.text[:50]!r}"
    )

```python
    async def search(self, query: SearchQuery) -> list[Evidence]:
        embedding = await self._client.embed(
            EmbeddingRequest(
                model=self.spec.embedding_model,
                texts=(query.text,),
                data_class=query.data_class,
                dimensions=self.spec.dimensions,
            )
        )
        hits = await self._store.search(
            self.spec, embedding.vectors[0], k=query.k, filters=query.filters or None
        )
        return [
            Evidence(
                text=hit.chunk.text,
                score=hit.score,
                rank=rank,
                provenance=Provenance(
                    source_key=self.key,
                    document_id=hit.chunk.doc_id,
                    chunk_id=hit.chunk.chunk_id,
                    locator=hit.chunk.locator(),
                    title=hit.document_title or None,
                    uri=hit.document_uri,
                    checksum=hit.chunk.checksum,
                    ingested_at=hit.ingested_at,
                    extra={"index_key": self.spec.index_key},
                ),
                metadata=dict(hit.document_metadata),
            )
            for rank, hit in enumerate(hits, start=1)
        ]
```

question : 'What is the first step of RAG?'

rang  score  document     localisateur                                 texte
   1  0.668  1001.0001    section « Introduction », caractères 0–56    'Retrieval finds documents.\n\nIt is the first step o'
   2  0.000  1001.0001    section « Abstract », caractères 0–33        'We study retrieval. It is useful.'
   3  0.000  1001.0001    section « Method », caractères 0–56          'We embed chunks.\n\nThen we search with cosine simil'
   4  0.000  1001.0002    section « Abstract », caractères 0–18        'We study chunking.'


### 4.2 Le prompt final : ce que le modèle voit vraiment

Le prompt versionné `prompts/chat_answer.md` reçoit les preuves numérotées avec leur provenance,
et deux règles : **citer** chaque affirmation par `[n]`, et **refuser** avec une phrase exacte
quand les extraits ne permettent pas de répondre. Le voici, rendu — c'est ce texte-là, et son
empreinte, qui partent dans la trace.

In [9]:
builder = AnswerBuilder(PromptRegistry.from_settings(settings))
request = builder.build_request(
    question, evidence, model=ModelRef.parse(CHAT_MODEL), data_class=DataClass.PUBLIC
)
system = request.system or ""
print("=== system prompt (début) ===")
print(system[:900] + ("\n[…]" if len(system) > 900 else ""))
print("\n=== message utilisateur ===")
print(request.turns[0].content)
print("\n=== métadonnées de trace ===")
print(request.metadata)

=== system prompt (début) ===
Tu réponds à des questions en t'appuyant UNIQUEMENT sur les extraits numérotés ci-dessous.

Règles :
1. Chaque affirmation de ta réponse cite l'extrait qui la soutient, avec son numéro entre crochets, par exemple [2]. Une phrase sans citation est interdite.
2. Ne cite jamais un numéro absent de la liste. N'invente aucune information absente des extraits.
3. Si les extraits ne permettent pas de répondre, réponds exactement, et seulement : Je ne peux pas répondre à cette question à partir des documents fournis.
4. Réponds dans la langue de la question, de façon concise.

Extraits :
[1] docs:qasper-nb2 › A Tiny Paper on Retrieval › section « Introduction », caractères 0–56
Retrieval finds documents.

It is the first step of RAG.

[2] docs:qasper-nb2 › A Tiny Paper on Retrieval › section « Abstract », caractères 0–33
We study retrieval. It is useful.

[3] docs:qasper-nb2 › A Tiny Paper on Re
[…]

=== message utilisateur ===
What is the first step of RAG?

=== 

### 4.3 La réponse, ses citations, et le refus

`ChatService.answer` enchaîne retrieval → prompt → génération → **parsing des citations** : les
`[n]` sont rapprochés des preuves (un numéro sans preuve est signalé comme « citation
inventée »), et la phrase-sentinelle est reconnue comme un refus. On pose ensuite une question
hors sujet : un système honnête doit refuser plutôt qu'inventer. En mode complet, c'est le vrai
modèle : observez s'il refuse — la phase 2 mesurera ce taux sur SQuAD 2.0.

In [10]:
service = ChatService(
    client=client, source=source, builder=builder, default_model=ModelRef.parse(CHAT_MODEL)
)

answer = await service.answer(ChatRequest(question=question, k=K))
print("réponse    :", answer.answer)
print("refus      :", answer.refused)
print("citations  :", [f"[{c.index}] {c.label}" for c in answer.citations])
print("inventées  :", answer.unknown_citations)
print(
    f"modèle     : {answer.model} — coût {answer.cost_usd} $ — {answer.latency_ms:.0f} ms — tentatives {answer.attempts}"
)

off_topic = await service.answer(ChatRequest(question="Quelle est la recette des crêpes ?", k=K))
print("\nhors sujet → refus :", off_topic.refused, "—", off_topic.answer[:120])

réponse    : D'après les extraits, la première étape est le retrieval [1]. Il précède la génération [2].
refus      : False
citations  : ['[1] docs:qasper-nb2 › A Tiny Paper on Retrieval › section « Introduction », caractères 0–56', '[2] docs:qasper-nb2 › A Tiny Paper on Retrieval › section « Abstract », caractères 0–33']
inventées  : []
modèle     : claude-sonnet-5 — coût 0.0017 $ — 12 ms — tentatives 1

hors sujet → refus : True — Je ne peux pas répondre à cette question à partir des documents fournis.


### 4.4 En flux : les preuves d'abord, les tokens ensuite, le bilan à la fin

Le même service, en SSE : le client reçoit les preuves **avant** le premier token (il peut
afficher les sources pendant que la réponse s'écrit), puis les tokens, puis le bilan avec les
citations parsées et le coût. C'est ce que `/chat/stream` envoie sur le réseau.

In [11]:
started = time.perf_counter()
async for event in service.stream(ChatRequest(question=question, k=2)):
    elapsed = (time.perf_counter() - started) * 1000
    if event.kind == "retrieval":
        print(f"+{elapsed:5.0f} ms  retrieval  {len(event.evidence or [])} preuves")
    elif event.kind == "token":
        print(f"+{elapsed:5.0f} ms  token      {event.text!r}")
    elif event.kind == "done" and event.answer is not None:
        print(
            f"+{elapsed:5.0f} ms  done       citations={[c.index for c in event.answer.citations]} coût={event.answer.cost_usd}"
        )

+    4 ms  retrieval  2 preuves
+   21 ms  token      'Le retrieval '
+   21 ms  token      'trouve les documents '
+   21 ms  token      '[1].'
+   24 ms  done       citations=[1] coût=0.0003


## Partie 5 — Les chiffres : trois modèles d'embeddings sur SciFact, contre la littérature

Le protocole est celui de **MTEB**, exécuté par sa bibliothèque de référence (`mteb`) : SciFact,
split test, 300 requêtes contre 5 183 abstracts, métrique **nDCG@10** (entre 0 et 1, récompense
les documents pertinents placés en tête des dix premiers). Les modèles ouverts tournent en local ;
le modèle d'API passe par **notre** `LLMClient.embed` grâce à un adaptateur (`ClientEncoder`) —
même gouvernance, même coût, mêmes traces qu'en production.

Les résultats sont archivés dans `data/results/embeddings_scifact.json` (produit par
`make bench-embeddings`, jamais par la CI) et confrontés aux scores **publiés** : la carte de
modèle de bge-base-en-v1.5, et BM25 du papier BEIR (0,665) — le repère à battre.

In [12]:
from llm_testbench.eval.embeddings_bench import ClientEncoder, run_scifact, save_report

show_source(ClientEncoder.encode)

results_path = ROOT / "data" / "results" / "embeddings_scifact.json"
if RUN_API_BENCH and USE_REAL_API:
    report_bench = run_scifact([EMBEDDING_MODEL], client=client, api_dimensions=None)
    save_report(report_bench, ROOT / "data" / "results" / "embeddings_scifact_api.json")
    print(report_bench.table())

if results_path.is_file():
    bench = load_report(results_path)
    print(
        f"\nrapport du {bench.generated_at:%Y-%m-%d} — tâche {bench.task}, métrique {bench.metric}\n"
    )
    print(bench.table())
    bm25, bm25_source = PUBLISHED_SCIFACT["BM25 (Anserini)"]
    print(f"\nrepère lexical : BM25 = {bm25:.3f} nDCG@10 ({bm25_source})")
    for score in bench.scores:
        verdict = "au-dessus" if score.ndcg_at_10 > (bm25 or 0) else "EN DESSOUS"
        print(f"  {score.model:<42} {verdict} de BM25 ({score.ndcg_at_10 - (bm25 or 0):+.3f})")
else:
    print("pas de rapport : lancez `make bench-embeddings` (modèles locaux ~5 min, API ~0,03 $)")

```python
    def encode(self, inputs: Iterable[dict[str, Any]], **_: Any) -> np.ndarray:
        """``inputs`` est un DataLoader de lots ``{"text": [...], ...}`` (protocole mteb)."""
        batches = [self._embed_texts(list(batch["text"])) for batch in inputs]
        return np.concatenate(batches, axis=0) if batches else np.zeros((0, 0), dtype=np.float32)
```


rapport du 2026-09-03 — tâche SciFact, métrique ndcg_at_10

modèle                                      nDCG@10   publié   écart   dim   durée      coût
sentence-transformers/all-MiniLM-L6-v2        0.645        —       —   384     20s     local
BAAI/bge-base-en-v1.5                         0.740    0.740  -0.000   768    120s     local

repère lexical : BM25 = 0.665 nDCG@10 (Thakur et al. 2021, BEIR, table 2)
  sentence-transformers/all-MiniLM-L6-v2     EN DESSOUS de BM25 (-0.020)
  BAAI/bge-base-en-v1.5                      au-dessus de BM25 (+0.075)


**Ce que dit le tableau** (rapport archivé le 2026-09-03, deux modèles locaux ; le modèle d'API
s'ajoute avec `make bench-embeddings` et une clé) :

- **bge-base-en-v1.5 reproduit exactement son score publié** (0,740 contre 0,7404 sur sa carte
  de modèle). Le protocole est donc bon : quand notre harness de la phase 2 réimplémentera
  nDCG@10, il devra retrouver ce chiffre — sinon c'est le harness qui est faux.
- **all-MiniLM-L6-v2, le modèle de tous les tutoriels, fait moins bien que BM25** (0,645 contre
  0,665). Un RAG « moderne » construit sur lui perd contre une recherche par mots-clés de 1994 sur
  ce jeu. C'est la raison d'être de la recherche hybride (phase 3) — et une bonne question
  d'entretien.
- L'écart entre les deux modèles ouverts (près de dix points) coûte 768 dimensions au lieu de
  384 et quatre fois le temps d'encodage : le prix se voit dans la colonne « durée » et se
  paiera dans la taille de l'index.

## Les pièges

1. **Deux modèles d'embeddings, un seul index** : la question doit être embarquée par le
   modèle qui a embarqué les chunks. Le registre d'index porte le modèle et sa dimension ; le
   store refuse un vecteur de mauvaise dimension.
2. **La troncature silencieuse** : au-delà de sa fenêtre, un modèle d'embeddings ignore la fin
   du chunk sans erreur. Le chunk de 4 000 caractères est indexé sur ses 2 000 premiers.
3. **Le chevauchement n'est pas gratuit** : 150 caractères de chevauchement sur des chunks de
   800, c'est ~19 % de tokens d'embeddings payés deux fois, et des doublons dans les résultats.
4. **Le score cosinus n'est pas une probabilité** : 0,45 peut être excellent avec un modèle et
   médiocre avec un autre ; seule la mesure (nDCG contre des jugements) dit ce qu'il vaut.
5. **Une citation inventée ressemble à une vraie** : `[7]` quand il n'y a que quatre preuves.
   Le parseur la signale (`unknown_citations`) ; la phase 2 en fera un taux.
6. **Le refus par phrase exacte est fragile** : un modèle qui refuse « à sa façon » n'est pas
   compté comme refus. Le parseur tolère accents, casse et ponctuation, pas les paraphrases —
   c'est un choix mesurable, pas une vérité.
7. **HNSW est approximatif** : il peut rater un voisin proche pour aller vite (`ef_search`).
   Sur 5 000 chunks c'est invisible ; à un million, c'est un paramètre à mesurer (phase 3).
8. **Le rapport de la CI n'est pas un chiffre** : en mode échantillon, l'embedder est simulé et
   deux articles sont indexés. Seul un rapport produit à la main, en mode complet, se publie.

## À vous de jouer

1. **`CHUNK_SIZE = 300` / `CHUNK_OVERLAP = 0`** (mode complet) — plus de chunks, plus petits :
   regardez le nombre de chunks, les scores de retrieval, et si la réponse cite encore juste.
   L'index est reconstruit automatiquement quand la définition change.
2. **`K = 1`** puis **`K = 10`** — trop peu de preuves : le modèle refuse ou invente ; trop :
   le prompt grossit (coût) et le bruit monte.
3. **`QUESTION_INDEX = 1`, `2`** — d'autres questions annotées du même article ; certaines n'ont
   pas de réponse dans le texte (QASPER en contient) : le système refuse-t-il ?
4. **`SHOW_PAPER = 1`** — un autre article, une autre structure de sections.
5. **`RUN_API_BENCH = True`** (mode complet, ~0,03 $) — ajoute text-embedding-3-small au tableau.

## Questions d'entretien

<details><summary><b>1. Comment choisissez-vous la taille des chunks ?</b></summary>

En mesurant : la taille est un paramètre de l'index, chaque valeur donne un index distinct dans
le registre, et le harness compare recall@k et nDCG sur un jeu à jugements (phase 3). Les
contraintes fixes : la fenêtre du modèle d'embeddings (troncature au-delà) et la fenêtre du
générateur (k chunks doivent tenir avec de la marge). L'intuition « un paragraphe » est un point
de départ, pas une réponse.
</details>

<details><summary><b>2. Pourquoi une table de vecteurs par index plutôt qu'une colonne dans la table des chunks ?</b></summary>

Parce qu'un index HNSW exige une dimension fixée et qu'on veut comparer des modèles de
dimensions différentes sur les mêmes chunks. Les chunks sont partagés (parsés une fois), les
vecteurs dépendent du modèle. Le registre relie chaque index à ce qui l'a produit : c'est le
lineage sans lequel un tableau de résultats n'est pas reproductible.
</details>

<details><summary><b>3. Un modèle d'embeddings « moderne » peut-il perdre contre BM25 ?</b></summary>

Oui, et ce notebook le montre : all-MiniLM-L6-v2 fait 0,645 sur SciFact contre 0,665 pour BM25.
Les petits modèles denses généralistes sont faibles sur le vocabulaire rare et les acronymes
d'un domaine ; BM25 les prend au mot. D'où la recherche hybride (dense + lexical, fusion RRF)
et le re-ranking, mesurés en phase 3.
</details>

<details><summary><b>4. Comment garantissez-vous qu'une réponse cite ses sources et n'invente pas ?</b></summary>

On ne garantit pas, on mesure. Le prompt impose une citation par affirmation et une phrase de
refus exacte ; le parseur rapproche chaque `[n]` d'une preuve réelle et signale les numéros
inventés ; la phase 2 mesure la fidélité (groundedness), l'exactitude des citations et les taux
de refus correct et abusif. Une garantie sans mesure est une promesse.
</details>

<details><summary><b>5. Que se passe-t-il quand on met à jour un document déjà indexé ?</b></summary>

Son empreinte change : le pipeline réécrit ses chunks, les anciens vecteurs partent en cascade,
seuls les nouveaux chunks (empreinte inconnue de l'index) sont ré-embarqués. Un document
identique est un no-op. La phase 3 ajoutera la détection de changement et la réindexation
incrémentale sur le corpus vivant.
</details>

## Ce qu'on n'a pas fait, et pourquoi

- **Pas de PDF** : QASPER fournit le texte intégral ; le parsing de vrais PDF (colonnes,
  tableaux, notes) est mesuré en phase 3 sur le corpus maison difficile à parser.
- **Pas de recherche hybride ni de re-ranking** : phase 3, derrière un flag, chaque technique
  mesurée contre cette baseline.
- **Pas de métrique de retrieval maison** : le harness (recall@k, MRR, nDCG) est la phase 2 ;
  ici, `mteb` fait référence — et fixe le chiffre que le harness devra retrouver.
- **Pas de cache sémantique en base** : le backend Postgres du cache arrive avec les premières
  requêtes répétées à mesurer (phase 5).
- **Modèle d'API absent du rapport archivé** : il demande une clé ; `make bench-embeddings`
  l'ajoute en deux minutes.

**La suite** : phase 2, l'évaluation — le harness qui produira, sur ces mêmes index, les chiffres
du tableau maître.

## Lexique récapitulatif

| Terme | Définition |
|---|---|
| **RAG** | Retrieval-augmented generation : chercher des passages, puis générer en les citant. |
| **Ingestion** | Préparer les documents pour la recherche : parsing, chunking, embeddings, index. |
| **Chunk** | Morceau de document indexé et retrouvé comme un tout. |
| **Chevauchement** | Texte partagé entre deux chunks consécutifs, pour ne pas couper une idée sans témoin. |
| **Embedding** | Vecteur représentant le sens d'un texte ; deux textes proches ont des vecteurs proches. |
| **Similarité cosinus** | Mesure d'angle entre deux vecteurs : 1 identique, 0 sans rapport. |
| **pgvector / HNSW** | Extension Postgres pour les vecteurs / index approximatif pour les chercher vite. |
| **Index (registre)** | Un jeu de vecteurs produit par un chunker et un modèle donnés, décrit dans la table `indexes`. |
| **Evidence / Provenance** | Un passage retrouvé, avec d'où il vient (source, document, section, positions, empreinte). |
| **Citation `[n]`** | Renvoi d'une affirmation vers la n-ième preuve du prompt. |
| **Phrase-sentinelle** | Réponse exacte qui signifie « je ne peux pas répondre » et que le parseur reconnaît. |
| **nDCG@10** | Métrique de retrieval (0 à 1) récompensant les documents pertinents placés en tête des dix premiers. |
| **MTEB / BEIR** | Le protocole standard de comparaison d'embeddings / la collection de jeux de retrieval dont SciFact fait partie. |
| **BM25** | Recherche par mots-clés classique ; la baseline lexicale à battre. |

In [13]:
# Ménage : en mode échantillon, l'index de démonstration ne survit pas au notebook.
if SAMPLE:
    await store.drop_index(INDEX_KEY)
    await store.delete_source(SOURCE_KEY)
    print("index et documents de démonstration supprimés")
await store.close()
print(f"spans émis pendant le notebook : {len(exporter.get_finished_spans())}")

index et documents de démonstration supprimés
spans émis pendant le notebook : 9
